Services/RequestServer/WebSocketImplCurl.cpp

```diff
@@ -0,0 +1,175 @@
+        return false;
+    };
+
+    set_option(CURLOPT_PRIVATE, reinterpret_cast<uintptr_t>(this) | 1);
+    set_option(CURLOPT_WS_OPTIONS, CURLWS_RAW_MODE);
+    set_option(CURLOPT_CONNECT_ONLY, 2); // WebSocket mode
+
+    // FIXME: Add a header function to validate the Sec-WebSocket headers that curl currently doesn't validate
+
+    auto const& url = info.url();
+    set_option(CURLOPT_URL, url.to_byte_string().characters());
+    set_option(CURLOPT_PORT, url.port_or_default());
+
+    if (auto root_certs = info.root_certificates_path(); root_certs.has_value())
+        set_option(CURLOPT_CAINFO, root_certs->characters());
+
+    auto const origin_header = ByteString::formatted("Origin: {}", info.origin());
+    curl_slist* curl_headers = curl_slist_append(nullptr, origin_header.characters());
+
+    for (auto const& [name, value] : info.headers().headers()) {
+        // curl will discard headers with empty values unless we pass the header name followed by a semicolon.
+        ByteString header_string;
+        if (value.is_empty())
+            header_string = ByteString::formatted("{};", name);
+        else
+            header_string = ByteString::formatted("{}: {}", name, value);
+        curl_headers = curl_slist_append(curl_headers, header_string.characters());
+    }
+
+    if (auto const& protocols = info.protocols(); !protocols.is_empty()) {
+        StringBuilder protocol_builder;
+        protocol_builder.append("Sec-WebSocket-Protocol: "sv);
+        protocol_builder.append(ByteString::join(","sv, protocols));
+        curl_headers = curl_slist_append(curl_headers, protocol_builder.to_byte_string().characters());
+    }
+
+    if (auto const& extensions = info.extensions(); !extensions.is_empty()) {
+        StringBuilder protocol_builder;
+        protocol_builder.append("Sec-WebSocket-Extensions: "sv);
+        protocol_builder.append(ByteString::join(","sv, extensions));
+        curl_headers = curl_slist_append(curl_headers, protocol_builder.to_byte_string().characters());
+    }
+
+    set_option(CURLOPT_HTTPHEADER, curl_headers);
+    m_curl_string_lists.append(curl_headers);
+
+    CURLMcode const err = curl_multi_add_handle(m_multi_handle, m_easy_handle);
+    VERIFY(err == CURLM_OK);
+}
+
+bool WebSocketImplCurl::can_read_line()
+{
+    VERIFY_NOT_REACHED();
+}
+
+ErrorOr<ByteBuffer> WebSocketImplCurl::read(int max_size)
+{
+    auto buffer = TRY(ByteBuffer::create_uninitialized(max_size));
+    auto const read_bytes = TRY(m_read_buffer.read_some(buffer));
+    return buffer.slice(0, read_bytes.size());
+}
+
+ErrorOr<ByteString> WebSocketImplCurl::read_line(size_t)
+{
+    VERIFY_NOT_REACHED();
+}
+
+bool WebSocketImplCurl::send(ReadonlyBytes bytes)
+{
+    size_t sent = 0;
+    CURLcode const result = curl_ws_send(m_easy_handle, bytes.data(), bytes.size(), &sent, 0, 0);
+    (void)sent;
+
+    return result == CURLE_OK;
+}
+
+bool WebSocketImplCurl::eof()
+{
+    return m_read_buffer.is_eof();
```

Comment:
I couldn't figure out how to do is_eof() without adding the extra read buffer in the way. More braincells appreciated to remove the read buffer and call curl_easy_recv from read(int) directly.